src/util/submit-addon.js

```diff
@@ -423,14 +455,23 @@ export async function signAddon({
     channel,
     savedUploadUuidPath,
   );
+  // if we have a source file we need to upload we patch after the create
+  const versionPatchData = versionSource
```

Comment:
Shouldn't we validate the source before attempting submission?

- That it looks like a zip file (opposed to non-existing file, a non-file like a directory, or a non-zip like a tar file)
- That it hasn't been submitted before, to ensure that the dev doesn't forget to update the zip file.

+ unit tests for these cases

Reply:
>  Shouldn't we validate the source before attempting submission?

Nothing else is really validated before submission - otherwise we'd be duplicating the validation that takes place in the API and opening ourselves up to two diverging standards.

> That it looks like a zip file (opposed to non-existing file, a non-file like a directory, or a non-zip like a tar file)

I don't think a non-existing file, or a directory would successfully be read by `fileFromSync`, so it'd just fail with an error at that point.  I've just looked and AMO actually accepts tar files too for sources, which kind of reinforces why we shouldn't try to do too much in `web-ext` locally because what is accepted could change at any time.  (And I should adjust the wording of the help text)

> That it hasn't been submitted before, to ensure that the dev doesn't forget to update the zip file.

I'm not sure this is practical - there are cases where the xpi could have changed but the source is still the same (e.g. bad build) - but if you want to log a follow-up issue we could spec out how it would work.

Reply:
> >  Shouldn't we validate the source before attempting submission?
> > That it looks like a zip file (opposed to non-existing file, a non-file like a directory, or a non-zip like a tar file)
> 
> I don't think a non-existing file, or a directory would successfully be read by `fileFromSync`, so it'd just fail with an error at that point.

My request is more about having a good developer experience than anything else. If the fileFromSync tool already emits useful error messages when a directory or non-existing file is passed, then the implementation doesn't need more. But in any case I'd like to see test coverage to ensure sensible long-term behavior.

> > That it hasn't been submitted before, to ensure that the dev doesn't forget to update the zip file.
> 
> I'm not sure this is practical - there are cases where the xpi could have changed but the source is still the same (e.g. bad build) - but if you want to log a follow-up issue we could spec out how it would work.

AMO requires unique version numbers. I would consider it quite unlikely for a source to be completely unchanged. I would consider it much more likely for someone to pass a path to a zip file to make the tool happy, and then forget to update the zip file later. Of course, when a reviewer catches this, they would reject the add-on on review, but IMO it would be worth detecting that mistake in advance.

Detection of source checks can be done similarly to how xpi duplicate checks are done now (i.e. comparing with the hash of the last successful submission). If you are concerned about blocking submission of the same source, another parameter to override the check can be added.

I'm willing to move this to a follow-up.

Reply:
https://github.com/mozilla/web-ext/issues/2892 for dealing with dup'd source files

Reply:
https://github.com/mozilla/web-ext/pull/2859/commits/26efa493111b7ce80e6b794ffc4704e9c9898ac0 adds tests and specific checks.